### ISSUE TITLE
Cancellation and Conditions

### ISSUE THREAD

**Issue by jshier** [danthorpe/Operations#183]
Up until recently I haven't been using conditions much. However, I'm now refactoring my network stack so that every operation has a condition that certain bits of information are set. So now I'm starting to run into this `assert` every so often, with no code changes from me:
```swift
assert(cancelled == false, "\(__FUNCTION__) was called on cancelled operation.")
```
This doesn't seem like something I can avoid in my own code. I'll investigate more but it's very intermittent, which seems to indicate a race between `cancelled` being set and this code being called:
```swift
case .Pending:
    // If the operation is cancelled, isReady should return true
    if cancelled {
        state = .Ready
        return true
    }
    if super.ready {
        return evaluateConditions()
    }
    // Until conditions have been evaluated, we're not ready
    return false
```
Any guidance?

**Comment by danthorpe** [comment]
Hi @jshier - thanks for this. I think it's the same known issue which I'm currently working on.

This bit of code is getting refactored. Essentially I'm making all of the properties (such as cancelled etc) thread safe.

**Comment by danthorpe** [comment]
For guidance - if you can work around it for now, that would be great. I hope to have this race condition fixed in development by end of day tomorrow.

**Comment by danthorpe** [comment]
Hi @jshier! Regarding this issue.. I think the issues you have experienced here might be related to the general KVO issues which have been investigated this week and are being tracked in #175, with some improvements merged in #180.

Would be interested to see if the changes in #180 have alleviated this issue for you at all. 

If not: I think the nest step is to clarify if it's a "cancelled" state race condition issue, or... something else I guess.

One thing which I do want to check on, is that we should be returning `"Cancelled"` (in addition to `"State"`) as a KVO path affecting `isReady` which is [a class function](https://github.com/danthorpe/Operations/blob/development/Sources/Core/Shared/Operation.swift#L78) inside `Operation`. If you're able to test this inside your fork that would be really helpful :)

**Comment by danthorpe** [comment]
@jshier I'll see if I can sort this out this week - would be great to fix this in 2.6.0.

**Comment by danthorpe** [comment]
Okay - so I've made this change in v2.6.0: see [here](https://github.com/danthorpe/Operations/commit/5ad6f63f9a47ae8bece413447ae5d812c931251c). I'm not actually sure if this will fix this issue, and I've not been able to write a unit test to reproduce this issue. However that change is something which should have been inlace already. If the cancelled state changes, so should the conditions readiness.

**Comment by jshier** [comment]
I haven't seen this in quite a while, so I think it was fixed in 2.6 or 2.7.

**Comment by danthorpe** [comment]
Thanks @jshier for getting back to me on this - gonna close it for now.
